auth/src/VSCodeAzureSessionProvider.ts

```diff
@@ -0,0 +1,241 @@
+        this.tenantSignInStatuses = newTenantSignInStatuses;
+        this.signInStatusValue = newSignInStatus;
+        if (signInStatusChanged || tenantsChanged) {
+            this.onSignInStatusChangeEmitter.fire(this.signInStatusValue);
+        }
+    }
+
+    /**
+     * Get the current Azure session, silently if possible.
+     * @returns The current Azure session, if available. If the user is not signed in, or there are no tenants,
+     * an error is thrown.
+     */
+    public async getAuthSession(tenantId: string, behavior: GetSessionBehavior, scopes?: string[]): Promise<AzureAuthenticationSession | undefined> {
+        await this.initializePromise;
+        if (this.signInStatusValue !== "SignedIn") {
+            throw new NotSignedInError();
+        }
+
+        const tenantSignInStatus = this.tenantSignInStatuses.find(s => s.tenant.tenantId === tenantId);
+        if (!tenantSignInStatus) {
+            throw new Error(`User does not have access to tenant ${tenantId}`);
+        }
+
+        // Get a session for a specific tenant.
+        scopes = scopes || [getDefaultScope(getConfiguredAzureEnv().resourceManagerEndpointUrl)];
+        const behaviourScenarios: Record<GetSessionBehavior, AuthScenario> = {
+            [GetSessionBehavior.Silent]: AuthScenario.GetSessionSilent,
+            [GetSessionBehavior.PromptIfRequired]: AuthScenario.GetSessionPrompt,
+        };
+
+        const session = await this.getArmSession(tenantId, behaviourScenarios[behavior], scopes);
+        tenantSignInStatus.isSignedIn = session !== undefined;
+
+        return session;
+    }
+
+    private async getArmSession(
+        tenantId: string,
+        authScenario: AuthScenario,
+        scopes: string[],
+    ): Promise<AzureAuthenticationSession | undefined> {
+        this.handleSessionChanges = false;
+        try {
+            scopes = addTenantIdScope(scopes, tenantId);
+
+            let options: AuthenticationGetSessionOptions;
+            let silentFirst = false;
+            switch (authScenario) {
+                case AuthScenario.Initialization:
+                case AuthScenario.GetSessionSilent:
+                    options = { createIfNone: false, clearSessionPreference: false, silent: true };
+                    break;
+                case AuthScenario.SignIn:
+                    options = { createIfNone: true, clearSessionPreference: true, silent: false };
+                    break;
+                case AuthScenario.GetSessionPrompt:
+                    // the 'createIfNone' option cannot be used with 'silent', but really we want both
+                    // flags here (i.e. create a session silently, but do create one if it doesn't exist).
+                    // To allow this, we first try to get a session silently.
+                    silentFirst = true;
+                    options = { createIfNone: true, clearSessionPreference: false, silent: false };
+                    break;
```

Comment:
I'm _pretty_ sure I tried that. However, I think it pops up a confirmation dialog if you enable `createIfNone`, which is what I was trying to avoid with the
`silent` option. I'd be happy to be proved wrong on that though - this is one of my least favourite parts of the code.

(It also might be a feature request on the VS Code side, but I'm not sure I've got a comprehensive idea of what the options/behaviour I'm asking for would be.)

Reply:
> (It also might be a feature request on the VS Code side, but I'm not sure I've got a comprehensive idea of what the options/behaviour I'm asking for would be.)

Yeah, it sounds like it could be a bug. Or at least the docstring for `createIfNone` is misleading.